src/software/amazon/ionschema/internal/TypeReference.kt

```diff
@@ -0,0 +1,98 @@
+package software.amazon.ionschema.internal
+
+import software.amazon.ion.IonStruct
+import software.amazon.ion.IonSymbol
+import software.amazon.ion.IonText
+import software.amazon.ion.IonValue
+import software.amazon.ionschema.InvalidSchemaException
+import software.amazon.ionschema.Schema
+
+/**
+ * Provides a factory method that translates an ISL type reference into a function
+ * that returns a Type instance.
+ *
+ * Types that can't be resolved yet are instantiated as [TypeReferenceDeferred] objects
+ * that are resolved by [SchemaImpl.resolveDeferredTypeReferences] prior to asserting
+ * that the schema is valid.
+ */
+internal class TypeReference private constructor() {
+    companion object {
+        fun create(ion: IonValue, schema: Schema, isField: Boolean = false): () -> TypeInternal {
+            var type: TypeInternal? = null
+            if (!ion.isNullValue) {
+                type = when (ion) {
+                    is IonStruct -> {
+                        val id = ion.get("id") as? IonText
+                        if (id != null) {
+                            // import
+                            val newSchema = schema.getSchemaSystem().loadSchema(id.stringValue())
+                            val typeName = ion.get("type") as IonSymbol
+                            newSchema.getType(typeName.stringValue()) as? TypeInternal
+
+                        } else {
+                            if (isField) {
+                                TypeImpl(ion, schema)
+                            } else {
+                                if (ion.size() == 1 && ion.get("type") != null) {
+                                    // elide inline types defined as "{ type: X }" to TypeImpl;
+                                    // this avoids creating a nested, redundant validation structure
+                                    TypeImpl(ion, schema)
+                                } else {
+                                    TypeInline(ion, schema)
+                                } as TypeInternal
+                            }
+                        }
+                    }
+
+                    is IonSymbol -> {
+                        val t = schema.getType(ion.stringValue())
+                        if (t != null) {
```

Comment:
favor when instead of long if/else chains

Reply:
this is now a single if/else, so i suspect the preference is to keep as-is (reserving `when` for cases with more than 2 conditions